pkg/controller/utils/statefulset.go

```diff
@@ -42,3 +55,87 @@ func CalculateStatefulSetSpecHash(statefulset *appsv1.StatefulSet) (string, erro
 
 	return sum, nil
 }
+
+// GetMyOrdinalID expects to be executed in the context of a pod which is a replica of a
+// StatefulSet. We thus expect the pod name to have an ordinal suffix ("-0", "-1", etc)
+// by which we can identify the replica. To discover this, we expect the pod name to be
+// present in an env var named HIVE_${controllerName}_POD_NAME.
+// The first return is the ordinal number.
+// The error return is non-nil if:
+// - we can't find the pod name env var.
+// - the last hyphen-delimited component of the pod name can't be parsed as an int.
+func GetMyOrdinalID(controllerName hivev1.ControllerName, logger log.FieldLogger) (int64, error) {
+	// For searchability, we expect to produce one of
+	// HIVE_CLUSTERSYNC_POD_NAME
+	// HIVE_MACHINEPOOL_POD_NAME
+	envKey := fmt.Sprintf("HIVE_%s_POD_NAME", strings.ToUpper(string(controllerName)))
+	logger.Debugf("Getting %s", envKey)
+	podname, found := os.LookupEnv(envKey)
+
+	if !found {
+		return -1, fmt.Errorf("environment variable %s not set", envKey)
+	}
+
+	logger.Debug("Setting ordinalID")
+	parts := strings.Split(podname, "-")
+	ordinalIDStr := parts[len(parts)-1]
+	ordinalID32, err := strconv.Atoi(ordinalIDStr)
+	if err != nil {
+		return -1, errors.Wrapf(err, "error converting ordinalID %q to int", ordinalIDStr)
+	}
+
+	logger.WithField("ordinalID", ordinalID32).Debug("ordinalID discovered")
+	return int64(ordinalID32), nil
+}
+
+func IsUIDAssignedToMe(c client.Client, deploymentName hivev1.DeploymentName, uid types.UID, myOrdinalID int64, logger log.FieldLogger) (bool, error) {
+	l := logger.WithField("deploymentName", deploymentName).WithField("myOrdinalID", myOrdinalID)
+	hiveNS := GetHiveNamespace()
+
+	sts := &appsv1.StatefulSet{}
+	err := c.Get(context.Background(), types.NamespacedName{Namespace: hiveNS, Name: string(deploymentName)}, sts)
+	if err != nil {
+		l.WithError(err).WithField("hiveNS", hiveNS).Error("error getting statefulset.")
+		return false, err
+	}
+
+	if sts.Spec.Replicas == nil {
+		return false, errors.New("sts.Spec.Replicas not set")
+	}
+
+	if sts.Status.CurrentReplicas != *sts.Spec.Replicas {
+		// This ensures that we don't have partial syncing which may make it seem like things are working.
+		// TODO: Remove this once https://issues.redhat.com/browse/CO-1268 is completed as this should no longer be needed.
```

Comment:
It still redirects fine, but Cluster Operator project has since been renamed to Hive.

Reply:
Yah; this comment was just ported over. I left it untouched for, um, "posterity" :P